examples-next/auth/keystone.ts

```diff
@@ -0,0 +1,62 @@
+import { config } from '@keystone-spike/keystone/schema';
+import { statelessSessions, withItemData } from '@keystone-spike/keystone/session';
+import { lists } from './schema';
+import { createAuth } from '@keystone-spike/auth';
+
+/**
+ * TODO: Implement validateItem. Would be invoked by the getItem() method in
+ * packages-next/auth/src/getExtendGraphQLSchema.ts
+ */
+
+// createAuth configures signin functionality based on the config below. Note this only implements
+// authentication, i.e signing in as an item using identity and secret fields in a list. Session
+// management and access control are controlled independently in the main keystone config.
+const { withAuth } = createAuth({
+  // This is the list that contains items people can sign in as
+  listKey: 'User',
+  // The identity field is typically a username or email address
+  identityField: 'email',
+  // The secret field must be a password type field
+  secretField: 'password',
+  // This ensures than an item is actually able to sign in
+  validateItem: ({ item }) => item.isEnabled,
+  // initFirstItem turns on the "First User" experience, which prompts you to create a new user
+  // when there are no items in the list yet
+  initFirstItem: {
+    // These fields are collected in the "Create First User" form
+    fields: ['name', 'email', 'password'],
+    // This is additional data that will be set when creating the first item
+    itemData: {
+      // We need to specify that isAdmin is true for the first item, so the user can access the
+      // Admin UI (see isAccessAllowed in the admin config below)
+      isAdmin: true,
+      // Only enabled users can sign in, so we need to set this as well
+      isEnabled: true,
+    },
+  },
+});
+
+// withAuth applies the signin functionality to the keystone config
+export default withAuth(
+  config({
+    name: 'KeystoneJS Auth Example',
+    db: {
+      adapter: 'mongoose',
+      url: 'mongodb://localhost/keystone-examples-next-auth',
+    },
+    admin: {
+      // Only allow users with the isAdmin checkbox set to access the Admin UI
+      isAccessAllowed: ({ session }: { session: any }) => !!session?.item?.isAdmin,
+    },
+    lists,
+    session: withItemData(
+      // Stateless sessions will store the listKey and itemId of the signed-in user in a cookie
+      statelessSessions({
```

Comment:
What does it mean to be "stateless"? Are these JWTs? How are they expired?

Reply:
Naming is hard 😛 

Stateless -- yes, they're effectively JWTs. There are a _lot_ of different approaches to expiry and we don't actually encode an opinion. Basically the data (in this case `{ itemId, listKey }` is stored in an encrypted cookie that we decrypt to hydrate the initial session value. (An alternative being, which needs its own example, using the cookie to store the ID of a session in a data store -- we've added a redis adapter for this so far in the new interfaces)

In this actual implementation though, `withItemData` is taking that cookie and querying the database, to hydrate the item. It will clear the session (i.e read it as `undefined`, not actually clear the cookie) if the item can't be found or doesn't validate.

So this is _basically_ working the same way Keystone does now, except each of these decisions is a composable piece of functionality. You could use the stateless sessions without the item hydration; or you could use a session database to store the itemId and listKey and use hydration; or any other combo.

We've tried to make how it currently works easy for people to use these high-level functions to compose as if it's config, but internally these are all independent things working together so it's much more flexible.

Reply:
That all makes sense.

I worry though about the security implications of have a forever-alive token. As long as the `itemID` is valid, that token is forever-valid and can never be expired.

Actual JWTs at least [have an `exp` key](https://tools.ietf.org/html/rfc7519#page-9) which forces the JWT processor to act as if the token is invalid after that expiry time.

That's not to mention the problem with JWTs in general with respect to state & expiration, which I outline with a strong sarcastic tone in [this Twitter thread](https://twitter.com/JessTelford/status/1210362515298779136):

> JWTs are a really great idea!
> ... Until they meet reality; then you realise they're useless.
> 
> JWTs are stateless! No more storing current sessions in a database! Hurray!
> Oh, you want to logout? Well, you can't invalidate a JWT, so you'll have to blacklist its ID.
> Where is that blacklist stored? Why, in a database of course!
> 
> JWTs can expire! Set a short TTL so even if one leaks, it wont by useful after 5 min.
> Your users hate logging in every 5 mins? Well, you can provide a `refreshToken`! That'll get you a new JWT!
> When does that `refreshToken` expire? Never, of course! Infinite JWTs for all!
> 
> The argument I've heard for this one is; The `refreshToken` is like a CSRF token; it's sent with every auth request and can only be used once.
> How do you know if it's been used? You store it on the `User` table of course!
> Oh, you want multi-device login... ¯\\\_(ツ)\_/¯
> 
> Hint: The answer is "A session table".

For a different perspective with a more balanced view (and less sarcastic tone), see @molomby's comment here: https://github.com/keystonejs/keystone/issues/170#issuecomment-420485917

At any rate, the change away from session based to token based is a large one that can have far reaching impacts on the security and auth model of Keystone apps.

Reply:
"Stateless" is a terrible/completely wrong name but +1 that naming is hard in this context.